cli/src/xcresult.rs

```diff
@@ -0,0 +1,345 @@
+use indexmap::indexmap;
+use quick_junit::{NonSuccessKind, Report, TestCase, TestCaseStatus, TestSuite, XmlString};
+use std::str;
+use std::{fs, process::Command};
+
+#[derive(Debug, Clone)]
+pub struct XCResultFile {
+    pub path: String,
+    results_obj: serde_json::Value,
+}
+
+const LEGACY_FLAG_MIN_VERSION: i32 = 70;
+
+fn xcrun(args: Vec<&str>) -> Result<String, anyhow::Error> {
+    if !cfg!(target_os = "macos") {
+        return Err(anyhow::anyhow!("xcrun is only available on macOS"));
+    }
+    let mut cmd = Command::new("xcrun");
+    let bin = cmd.args(args);
+    let output = bin.output().map_err(|_| {
+        anyhow::anyhow!("failed to run xcrun -- please make sure you have xcode installed")
+    })?;
+    let result = String::from_utf8(output.stdout)
+        .map_err(|_| anyhow::anyhow!("got non UTF-8 data from xcrun output"))?;
+    Ok(result)
+}
+
+fn xcrun_version() -> Result<i32, anyhow::Error> {
+    let version_raw = xcrun(vec!["--version"])?;
+    // regex to match version where the output looks like xcrun version 70.
+    let re = regex::Regex::new(r"xcrun version (\d+)")?;
+    if let Some(capture_group) = re.captures(&version_raw.to_string()) {
+        if let Some(version) = capture_group.get(1) {
+            return Ok(version.as_str().parse::<i32>().unwrap_or(0));
+        }
+    }
+    Err(anyhow::anyhow!("failed to parse xcrun version"))
+}
+
+fn xcresulttool(
+    path: &str,
+    options: Option<Vec<&str>>,
+) -> Result<serde_json::Value, anyhow::Error> {
+    let mut base_args = vec!["xcresulttool", "get", "--path", path, "--format", "json"];
+    let version = xcrun_version()?;
+    if version >= LEGACY_FLAG_MIN_VERSION {
+        base_args.push("--legacy");
+    }
+    if let Some(val) = options {
+        base_args.extend(val);
+    }
+    let output = xcrun(base_args)?;
+    serde_json::from_str(&output)
+        .map_err(|_| anyhow::anyhow!("failed to parse json from xcrun output"))
+}
+
+impl XCResultFile {
+    pub fn new(path: String) -> Result<XCResultFile, anyhow::Error> {
+        let binding = fs::canonicalize(path.clone())
+            .map_err(|_| anyhow::anyhow!("failed to get absolute path"))?;
+        let absolute_path = binding.to_str().unwrap_or("");
+        let results_obj = xcresulttool(absolute_path, None)?;
+        Ok(XCResultFile {
+            path: absolute_path.to_string(),
+            results_obj,
+        })
+    }
+
+    fn find_tests(&self, id: &str) -> Result<serde_json::Value, anyhow::Error> {
+        xcresulttool(&self.path, Some(vec!["--id", id]))
+    }
+
+    fn generate_id(&self, raw_id: &str) -> String {
+        uuid::Uuid::new_v5(&uuid::Uuid::NAMESPACE_URL, raw_id.as_bytes()).to_string()
+    }
```

Comment:
Without it there's a warning:
```
the candidate is defined in an impl for the type `XCResultFile` (rustc E0599)
```

Reply:
Weird, that's not the error message I see. Either way, the problem is likely because if the method doesn't accept `self` in some shape or form, you much call the method statically like so: `XCResultFile::generate_id(raw_id)` instead of ~`self.generate_id(raw_id)`~